File: src/features/progression/xp.ts
/**
 * XP curve and helpers (PLAN.txt §6.10). Pure and deterministic.
 *
 * Level represents activity/account progression (distinct from Hunter rank,
 * which represents demonstrated performance). Economy constants live here, not
 * in the UI, so the curve can be tuned in one place.
 */

/** XP required to advance FROM the given level to the next: round(100 * level^1.5). */
export function getXpRequiredForLevel(level: number): number {
  const l = Math.max(1, Math.floor(level));
  return Math.round(100 * Math.pow(l, 1.5));
}

export interface XpProgress {
  /** XP earned toward the current level (clamped to [0, required]). */
  current: number;
  /** XP needed to reach the next level. */
  required: number;
  /** current / required, in [0, 1]. */
  fraction: number;
}

/** Progress of `currentXp` within `level`. */
export function xpProgress(currentXp: number, level: number): XpProgress {
  const required = getXpRequiredForLevel(level);
  const current = Math.max(0, Math.min(currentXp, required));
  return { current, required, fraction: required > 0 ? current / required : 0 };
}

File: android-native/app/src/main/java/com/fitnessrpg/app/domain/progression/Xp.kt
package com.fitnessrpg.app.domain.progression

import kotlin.math.pow
import kotlin.math.roundToLong

/**
 * XP curve and helpers (PLAN.txt §6.10). Pure and deterministic. Level tracks
 * activity/account progression (distinct from Hunter rank).
 */

/** XP required to advance FROM the given level to the next: round(100 * level^1.5). */
fun getXpRequiredForLevel(level: Int): Int {
    val l = maxOf(1, level)
    return (100 * l.toDouble().pow(1.5)).roundToLong().toInt()
}

data class XpProgress(
    /** XP earned toward the current level (clamped to [0, required]). */
    val current: Int,
    /** XP needed to reach the next level. */
    val required: Int,
    /** current / required, in [0, 1]. */
    val fraction: Double,
)

/** Progress of [currentXp] within [level]. */
fun xpProgress(currentXp: Int, level: Int): XpProgress {
    val required = getXpRequiredForLevel(level)
    val current = maxOf(0, minOf(currentXp, required))
    return XpProgress(current, required, if (required > 0) current.toDouble() / required else 0.0)
}
